index.js

```diff
@@ -1,50 +1,144 @@
-            res
-                .status(500)
-                .json({ 
-                    error: "The posts information could not be retrieved." 
-                })
+
+	db.find()
+		.then(posts => {
+			res.status(200).json({ posts });
+		})
+		.catch(() => {
+			res.status(500).json({
+				error: 'The posts information could not be retrieved.',
+			});
         });
+        
 });
 
+// gets one post by ID
 server.get('/api/posts/:id', (req, res) => {
-    db.findById(req.params.id)
+
+	db.findById(req.params.id)
+		.then(post => {
+			if (post.length) {
+				res.status(200).json(post);
+			} else {
+				res.status(404).json({
+					message: 'The post with the specified ID does not exist.',
+				});
+			}
+		})
+		.catch(err => {
+			res.status(500).json({
+                message: 'The post information could not be retrieved.',
+                error: err
+			});
+        });
+        
+});
+
+// adds new post
+server.post('/api/posts', (req, res) => {
+    
+    const post = req.body;
+    
+    if (post.title && post.contents) {
+        db.insert(post)
+            .then(response => {
+                db.findById(response.id)
+                    .then(newPost => {
+                        res.status(201).json(newPost);
+                    })
+            })
+            .catch(err =>
+                res.status(500).json({
+                    message: 'There was an error while saving the post to the database',
+                    error: err
+                })
+            );
+    } else {
+        res.status(500).json({
+            message: 'Please provide title and contents for the post.'
+        })
+    }
+
+});
+
+// deletes post by ID
+server.delete('/api/posts/:id', (req, res) => {
+
+    const { id } = req.params;
+
+    db.findById(id)
         .then(post => {
-            if (post) {
-                res
-                    .status(200)
-                    .json({ post })
+            if (post.length) {
+                db.remove(id).then(() => {
+                    res.status(200).json(post);
+                })
             } else {
                 res
                     .status(404)
-                    .json({ 
-                        message: "The post with the specified ID does not exist." 
+                    .json({
+                        message: 'The post with the specified ID does not exist.'
                     })
             }
         })
-        .catch(error => {
-            res
-                .status(500)
-                .json({
-                    error: "The post information could not be retrieved."
-                })
-        });
+        .catch(err => res.status(500).json({
+            message: 'The post could not be removed.',
+            error: err
+        }));
+
+});
+
+// modifies post by ID
+server.put('/api/posts/:id', (req, res) => {
+
+    const { id } = req.params;
+    const { body } = req;
+
+    if (body.title && body.contents) {
+        db.findById(id)
```

Comment:
Here's your `put` code altered to use async await:
```javascript
server.put('/api/posts/:id', async (req, res) => {
  const { id } = req.params;
  const { body } = req;
  const { title, contents } = body;

  if (title && contents) {
    try {
      const getPost = () => await db.findById(id);
      const post = getPost();
      if (post.length) {
        const count = await db.update(id, body);
        if (count) {
          const updatedPost = getPost();
          res.status(200).json(updatedPost);
        } else {
          res.status(500).json({
            message: 'Something went wrong with modifying the post.'
          })
        }
      } else {
        res.status(404).json({
          message: 'The post with the specified ID does not exist.'
        })
      }
    }
    catch(error) {
      res.status(500).json({
        message: 'The post information could not be modified.',
        error
      })
    }
  } else {
    res.status(400).json({
      message: 'Please provide title and contents for the post.'
    })
  }
});
```

Reply:
and here's the same code with the nesting eliminated without anything too convoluted being done:
```javascript
server.put('/api/posts/:id', async (req, res) => {
  const { id } = req.params;
  const { body } = req;
  const { title, contents } = body;

  if (title && contents) {
    const getPost = () => await db.findById(id);
    try {
      let post = getPost(), count;
      if (post.length) {
        count = await db.update(id, body);
      } else {
        res.status(404).json({
          message: 'The post with the specified ID does not exist.'
        })
      }
      if (count) {
        post = getPost();
        res.status(200).json(post);
      } else {
        res.status(500).json({
          message: 'Something went wrong with modifying the post.'
        })
      }
    }
    catch(error) {
      res.status(500).json({
        message: 'The post information could not be modified.',
        error
      })
    }
  } else {
    res.status(400).json({
      message: 'Please provide title and contents for the post.'
    })
  }
});
```

Reply:
or, alternatively, getting rid of some of the variables that aren't really needed:
```javascript
server.put('/api/posts/:id', async (req, res) => {
  const { id } = req.params;
  const { title, contents } = req.body;

  if (title && contents) {
    const getPost = () => await db.findById(id);
    try {
      let count = 0;
      if (getPost().length) {
        count = await db.update(id, { title, contents });
      } else {
        res.status(404).json({
          message: 'The post with the specified ID does not exist.'
        })
      }
      if (count) {
        res.status(200).json(getPost());
      } else {
        res.status(500).json({
          message: 'Something went wrong with modifying the post.'
        })
      }
    }
    catch(error) {
      res.status(500).json({
        message: 'The post information could not be modified.',
        error
      })
    }
  } else {
    res.status(400).json({
      message: 'Please provide title and contents for the post.'
    })
  }
});
```

Reply:
You could also make helper functions pretty easily. For example:
```javascript
const sendRes = res => code => json => res.status(code).json(json);
// ... inside a route
  const sendServerError = sendRes(res)(500);
  // ... the catch block after a try block
    catch(error) {
      sendServerError({
        message: 'The post information could not be modified.',
        error
      })
    }
```
For a lot of similar data, you can end up making something like this so that you don't have to rewrite very similar endpoints.